### ISSUE TITLE
Inerconnection between Pods

### ISSUE THREAD

**Issue by HippyRock** [Azure/AKS#1034]
**What happened**:
I have two simple pods WEB and MONGO that should talk to each other. I have two Linux nodes in the cluster. Cluster is private and services are exposed with type: LoadBalancer and metadata->annotations->service.beta.kubernetes.io/azure-load-balancer-internal: "true" flag. 
So if both pods on the same node then WEB pod gets connected to MONGO successfully for by mongo's service external IP. But if they are on the different nodes then connection doesn't work. I also can connect successfully to mongo's external IP from my jump box outside of cluster.
Also some reason I can't connect between pods by service name (either way from same node or different). I tried to specify FQDN too: mongodb://mongo.default.svc.cluster.local:27017.

**What you expected to happen**:
1. Interconnection between pods should work regardless if pods reside on single of different nodes.
2. Interconnection between pods should work by service names.

**Anything else we need to know?**:
Pods don't see internet some reason via kube-dns:
```
> kubectl get svc kube-dns -n kube-system
> NAME       TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)         AGE
> kube-dns   ClusterIP   10.0.0.10    <none>        53/UDP,53/TCP   47d
```
From node:
```
$ dig google.com @10.0.0.10 +short
; <<>> DiG 9.10.3-P4-Ubuntu <<>> google.com @10.0.0.10 +short
;; global options: +cmd
;; connection timed out; no servers could be reached
azureuser@k8s-linuxpool1-29716288-0:~$ dig google.com

; <<>> DiG 9.10.3-P4-Ubuntu <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 26879
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4000
;; QUESTION SECTION:
;google.com.                    IN      A

;; ANSWER SECTION:
google.com.             300     IN      A       172.217.5.238

;; Query time: 32 msec
;; SERVER: 10.245.191.181#53(10.245.191.181)
;; WHEN: Mon Jun 03 18:49:01 UTC 2019
;; MSG SIZE  rcvd: 55
```
From Pod1:
```
# apt-get update
Err:1 http://security.ubuntu.com/ubuntu xenial-security InRelease
  Temporary failure resolving 'security.ubuntu.com'
…

# cat /etc/resolv.conf
nameserver 10.0.0.10
search default.svc.cluster.local svc.cluster.local cluster.local reddog.microsoft.com edeazclabs.va.gov
options ndots:5
```
From Pod2:
```
> kubectl exec -ti busybox -- nslookup kubernetes.default
Server:    10.0.0.10
Address 1: 10.0.0.10

nslookup: can't resolve 'kubernetes.default'
command terminated with exit code 1
```
**Environment**:
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.0", GitCommit:"641856db18352033a0d96dbc99153fa3b27298e5", GitTreeState:"clean", BuildDate:"2019-03-25T15:53:57Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:02:58Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}

```

- Size of cluster: 2 Linux nodes, 2 windows node (hybrid cluster)
- General description of workloads in the cluster: HTTP microservices

Thanks, Denis.

**Comment by Senai-T** [comment]
is there a service with ClusterIP assigned to the pods? sometimes a clusterIP is needed for pod to pod communication...that's to say that containers inside the pod have a virtual subnet they can communicate on but clusterwide communication is facilitated by a clusterIP so give this a try:
apiVersion: v1
kind: Service
metadata:
  name: mongoDB
spec:
  selector:
    app: mongoDB
---
apiVersion: v1
kind: Service
metadata:
  name: web
spec:
  selector:
    app: web
  ports:
  - protocol: TCP
    port: 80

**Comment by Senai-T** [comment]
are they on the same subnet within the vnet? Do the both pods have a service with a type: ClusterIP?

**Comment by HippyRock** [comment]
All Linux pods are in the same subnet that is specified in cluster template. The services have type LoadBalancer:

```
apiVersion: v1
kind: Service
metadata:
  name: web
  annotations:
    service.beta.kubernetes.io/azure-load-balancer-internal: "true"
spec:
  type: LoadBalancer
  ports:
  - name: "80"
    port: 80
    targetPort: 80
  selector:
    app: web
```

**Comment by Senai-T** [comment]
generally speaking, internal loadbalancers are best for connecting workloads outside your cluster (BUT inside your VNET) to services inside the cluster (sounds confusing right!). For this particular situation I'd recommend assigning both pods a clusterIP, this will expose them cluster-wide (modify as needed) :
web clusterIP:
apiVersion: v1
kind: Service
metadata:
  name: web
spec:
  type: ClusterIP
  ports:
  - name: "80"
    port: 80
    targetPort: 80
  selector:
    app: web 
mongoDB clusterIP:
apiVersion: v1
kind: Service
metadata:
  name: mongoDB
spec:
  type: ClusterIP
  ports:
  - name: "MongoDB-ClusterIP"
    port: 27017
  selector:
    app: mongoDB

**Comment by Senai-T** [comment]
generally speaking, internal loadbalancers are best for connecting workloads outside your cluster (BUT inside your VNET) to services inside the cluster (sounds confusing right!). For this particular situation I'd recommend assigning both pods a clusterIP, this will expose them cluster-wide (modify as needed) :
p.s. idk why my dashes are coming out as bullet points :(
web clusterIP:
apiVersion: v1
kind: Service
metadata:
  name: web
spec:
  type: ClusterIP
  ports:
   - name: "80"
     port: 80
     targetPort: 80
  selector:
     app: web 
mongoDB clusterIP:
apiVersion: v1
kind: Service
metadata:
  name: mongoDB
spec:
  type: ClusterIP
  ports:
   - name: "MongoDB-ClusterIP"
     port: 27017
  selector:
    app: mongoDB

**Comment by HippyRock** [comment]
Thanks, I tried ClusterIP type and result is the same as for LoadBalancer, connection works only when both pods are on the same node... 
In my test environment, I have nginx running on node 1 and busybox1 and busybox2 running on different nodes respectively. 
nginx has ClusterIP=10.0.79.54 and ExternalIP=10.245.194.200

```
> kubectl exec -it busybox -- wget -UseBasicParsing http://10.0.79.54:80
Connecting to 10.0.79.54:80 (10.0.79.54:80)
wget: can't connect to remote host (10.0.79.54): Connection timed out
command terminated with exit code 1
> kubectl exec -it busybox2 -- wget -UseBasicParsing http://10.0.79.54:80
Connecting to 10.0.79.54:80 (10.0.79.54:80)
wget: can't open 'index.html': File exists
command terminated with exit code 1
> kubectl exec -it busybox -- wget -UseBasicParsing http://10.245.194.200:80
Connecting to 10.245.194.200:80 (10.245.194.200:80)
wget: can't connect to remote host (10.245.194.200): Connection timed out
E0605 13:47:47.671816    5820 v2.go:105] EOF
command terminated with exit code 1
> kubectl exec -it busybox2 -- wget -UseBasicParsing http://10.245.194.200:80
Connecting to 10.245.194.200:80 (10.245.194.200:80)
wget: can't open 'index.html': File exists
command terminated with exit code 1

```
PS: I was not able to access ClusterIP from outside cluster from my jump box, not sure if that is by design.

**Comment by Senai-T** [comment]
yes the clusterIP is limited to the cluster only. Could you provide me with the yaml of the pods/deployments I'm going to attempt to recreate them and see what's what.

**Comment by HippyRock** [comment]
Ok, I have two Linux nodes with name=linux1 and linux2 labels respectively. And run kubectl apply for the following:

NGINX:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 1
  template:
    metadata:
      name: nginx
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - containerPort: 80
      nodeSelector:
        name: linux1
  selector:
    matchLabels:
      app: nginx
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  annotations:
    service.beta.kubernetes.io/azure-load-balancer-internal: "true"
spec:
  type: LoadBalancer
  ports:
  - protocol: TCP
    name: "80"
    port: 80
    targetPort: 80
  selector:
    app: nginx
```

and BUSYBOX
```
apiVersion: v1
kind: Pod
metadata:
  name: busybox2
  namespace: default
spec:
  containers:
  - name: busybox
    image: busybox:1.28
    command:
      - sleep
      - "3600"
    imagePullPolicy: IfNotPresent
  restartPolicy: Always
  nodeSelector:
    name: linux2
```

After everything up and running, you can try to connect to busybox pod and try to connect NGINX via ClusterIP or External IP by using wget, if connection works you should see something about index.html in the output:

```
kubectl exec -it busybox -- wget -UseBasicParsing http://10.0.79.54:80
Connecting to 10.245.194.200:80 (10.245.194.200:80)
wget: can't open 'index.html': File exists
```

**Comment by Senai-T** [comment]
so here is what I did, I threw the nginx pod on one node and two busybox pods on another node and then attached a ClusterIP to the nginx pod and then attempted to ping it using the cluster IP, it simply did not work but what DID work was using was using the endpoint (kubectl get ep)...I think this has to do with the networking structure of Kubenet  networking structure and the way IP address are siphoned from the node. 
So in conclusion, add a cluster IP to nginx as well as the pods (simply add a label to the pod descriptor and reference the label name in the clusterIP service descriptor), run kubectl get ep, grab the ip address provided and run kubectl exec -it busybox -- wget -UseBasicParsing http:{ep-address).
I will keep the deployments running on my end incase something doesn't work.

**Comment by HippyRock** [comment]
So, for you access to endpoint works from both nodes? I tried that and result is the same, no access to endpoint from pod that is on the other node...

I tried to troubleshoot DNS issues per this [guide ](https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/)and resolving "kubernetes.default" doesn't work from pod or node:
```
> kubectl exec -ti busybox -- nslookup kubernetes.default
Server:    10.0.0.10
Address 1: 10.0.0.10

nslookup: can't resolve 'kubernetes.default'
command terminated with exit code 1
```
Also coreDns pod doesn't receive any routing requests (tested per instruction from the same guide).

Cluster is created with the following profile:
```
                       "orchestratorProfile":  {
                                                   "orchestratorType":  "Kubernetes",
                                                   "orchestratorRelease": "1.14",
                                                   "orchestratorVersion": "1.14.1",
                                                   
                                                   "kubernetesConfig": {
                                                    "enableRbac": false,
                                                    "serviceCidr": "10.0.0.0/16",
                                                    "clusterSubnet": "192.168.0.0/16",
                                                    "dnsServiceIP": "10.0.0.10",
                                                    "networkPlugin": "kubenet",

                                                    "privateCluster": {
                                                        "enabled": true
                                                    },
                                                    "kubeletConfig": {
                                        
                                                      "--max-pods": "30"
                                                    }   
                                                  }
                                               }
```

I wonder if it has to do something with clusterSubnet? I noticed that endpoints IPs are allocated from that subnet (192.168.0.0/16), per information from [here](https://github.com/Azure/aks-engine/blob/master/docs/topics/clusterdefinitions.md), clusterSubnet must be in must be in the VNET address space.

PS: as a side note, codeDns is used, but its pod some reason has  k8s-app=kube-dns label. Which is confusing, but maybe it doesn't affect anything...
```
> kubectl get pods --namespace=kube-system -l k8s-app=kube-dns
NAME                       READY   STATUS    RESTARTS   AGE
coredns-7bc69b7975-84hrl   1/1     Running   5          49d
```

**Comment by jnoller** [comment]
First, thank you for reaching out to the AKS team. 

AKS Github issues must be generally re-creatable (e.g here is the YAML/settings/packages/etc installed, can recreate the issue on existing clusters as well as newly created clusters). Unfortunately, this issue is cluster-specific.

For cluster specific issues (those that will require support and on-call staff to access you Subscription and clusters) we ask that you file an Azure Technical Support ticket with the identified issue, rec-creation steps if available, and link to this github thread.

**We can not ask** for the access, logs and other sensitive information on github, or ask you to email/slack/etc this information to us. Our portal ticketing system will protect the information including Subscription ids and credentials. If we discover as part of the root cause that this is an AKS or upstream bug, we will mark this issue as a known issue and ensure it is on the Product backlog as a repair item.

This will allow us to root-cause the issue and ensure it’s not specific to the environment or other configuration level issue - such as NSGs, custom network devices, etc.

Jesse
